src/adapter/src/catalog/migrate.rs

```diff
@@ -182,6 +209,576 @@ pub(crate) async fn migrate(
+                            .iter()
+                            .find(|(_, statement)| match statement {
+                                Statement::CreateSource(stmt) => stmt.name == name,
+                                _ => false,
+                            })
+                            .expect("source must exist");
+                        RawItemName::Id(source_item.id.to_string(), name, None)
+                    }
+                    RawItemName::Id(..) => raw_source_name,
+                };
+
+                // The external reference is a `with_option` on subsource statements but is a
+                // separate field on table statements.
+                let mut i = 0;
+                let external_reference = loop {
+                    if i >= with_options.len() {
+                        panic!("subsource must have an external reference");
+                    }
+                    if with_options[i].name == CreateSubsourceOptionName::ExternalReference {
+                        let option = with_options.remove(i);
+                        match option.value {
+                            Some(WithOptionValue::UnresolvedItemName(name)) => break name,
+                            _ => unreachable!("external reference must be an unresolved item name"),
+                        };
+                    } else {
+                        i += 1;
+                    }
+                };
+                let with_options = with_options
+                    .into_iter()
+                    .map(|option| {
+                        match option.name {
+                            CreateSubsourceOptionName::Details => TableFromSourceOption {
+                                name: TableFromSourceOptionName::Details,
+                                // The `details` option on both subsources and tables is identical, using the same
+                                // ProtoSourceExportStatementDetails serialized value.
+                                value: option.value,
+                            },
+                            CreateSubsourceOptionName::TextColumns => TableFromSourceOption {
+                                name: TableFromSourceOptionName::TextColumns,
+                                value: option.value,
+                            },
+                            CreateSubsourceOptionName::ExcludeColumns => TableFromSourceOption {
+                                name: TableFromSourceOptionName::ExcludeColumns,
+                                value: option.value,
+                            },
+                            CreateSubsourceOptionName::Progress => {
+                                panic!("progress option should not exist on this subsource")
+                            }
+                            CreateSubsourceOptionName::ExternalReference => {
+                                // This option is handled separately above.
```

Comment:
You can probably just move this comment into the `unreachable!()` call. i.e.

`unreachable!("This option is handled separately above.")`